codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -495,20 +518,22 @@ private void incrementPendingBytes(int numBytes) {
              * Note: this does not flush the {@link ChannelHandlerContext}.
              */
             int write(int allowedBytes) {
-                int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
                 try {
+                    pendingWriteQueue.remove();
```

Comment:
@nmittler - I think the response was for me (not @buchgr :) )...so I'll field the response

> The flow controller must protect its queue state in all cases.

I agree about managing the queue in general but I think the meaning of "protect" I may need a bit more details on.  Seems more straightforward if the interface is "state is assumed not to change until after flow controller has been notified of write completion...or else error".  This way for our implementation items can stay in the queue until they have been serviced, and if they are still in the queue when we are clearing it then it is OK to consider it an error.  Let me list out the sequence of events and you can poke holes in it.

1. Flow controller invokes payload.write().
2. "Encoder" (FlowControlled class that lives in the encoder) writes data.
3. Repeat 1 and 2 until no more to write
4. "Encoder" tells listener (which would be the flow controller) that it is done writing.
5. Flow controller cleans up its queue
6. "Encoder" continues on and cleans up if necessary (changes state via lifecycle manager if EOS).